File: javascript/No.0206-Reverse-Linked-List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
    //双指针，分别是当前current和pre,其中cur在后，pre在前，关系应该是pre=cur.next
    //两个指针的关系是同向前进
    let current = head;
    //如果当前node为空
    if (!current) {
        return current;
    }
    //这就成为了pre=cur.next的关系
    let curPre = head.next;
    //当前节点的next节点为null节点
    current.next = null;
    //只要先行节点不为空，就可以走下去。
    while (curPre) {
        // 存下先行节点
        let tmp = curPre.next;
        // 先行节点的下一个设置为当前节点,类似于调转链表箭头的操作
        curPre.next = current;
        //当前节点设置为先行节点，等于交换，也是调转链表箭头的操作
        current = curPre;
        //类似于给curPre先行指针+1
        curPre = tmp;
    }
    return current;

};


function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

var a1 = new ListNode(1, 2);
var a2 = new ListNode(0, a1);
var x = reverseList(a2)

console.log(a2);
console.log(x)
;

File: python/No0206_Reverse_Linked_List.py
# Given the head of a singly linked list, reverse the list, and return the
# reversed list.
#
#
#  Example 1:
#
#
# Input: head = [1,2,3,4,5]
# Output: [5,4,3,2,1]
#
#
#  Example 2:
#
#
# Input: head = [1,2]
# Output: [2,1]
#
#
#  Example 3:
#
#
# Input: head = []
# Output: []
#
#
#
#  Constraints:
#
#
#  The number of nodes in the list is the range [0, 5000].
#  -5000 <= Node.val <= 5000
#
#
#
#  Follow up: A linked list can be reversed either iteratively or recursively.
# Could you implement both?
#  Related Topics Linked List Recursion 👍 11865 👎 206


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for singly-linked list.
from typing import Optional


class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next


class Solution:
  def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
    dummy = ListNode(0)
    dummy.next = head

    while head is not None and head.next is not None:
      # 将位于列表首位的dummy的next取出来
      dnext = dummy.next
      # 将位于循环中的head的next取出来
      hnext = head.next

      # 交换开始
      # 首位换位置
      dummy.next = hnext
      # 循环中的位置替换成下一个
      head.next = hnext.next
      # 原来下一个的位置，替换为dummy.next，也就是当前循环cur
      hnext.next = dnext
    return dummy.next

# leetcode submit region end(Prohibit modification and deletion)
